planner/core/common_plans.go

```diff
@@ -706,18 +707,43 @@ func (e *Execute) rebuildRange(p Plan) error {
 			// TODO: relocate the partition after rebuilding range to make PlanCache support PointGet
 			return errors.New("point get for partition table can not use plan cache")
 		}
-		if x.HandleParam != nil {
-			var iv int64
-			iv, err = x.HandleParam.Datum.ToInt64(sc)
+		if x.HandleConstant != nil {
+			val, err := x.HandleConstant.Eval(chunk.Row{})
+			if err != nil {
+				return err
+			}
+			dVal, err := val.ConvertTo(sc, x.handleFieldType)
+			if err != nil {
+				return err
+			}
+			// The converted result must be same as original datum.
+			cmp, err := dVal.Compare(sc, &val, collate.GetCollator(x.handleFieldType.Collate))
+			if err != nil || cmp != 0 {
+				return errors.New("The parameter for point get can not use plan cache, because the parameter has changed after the covert")
+			}
+			iv, err := dVal.ToInt64(sc)
 			if err != nil {
 				return err
 			}
 			x.Handle = kv.IntHandle(iv)
 			return nil
 		}
-		for i, param := range x.IndexValueParams {
+		for i, param := range x.IndexConstants {
 			if param != nil {
-				x.IndexValues[i] = param.Datum
+				val, err := param.Eval(chunk.Row{})
+				if err != nil {
+					return err
+				}
+				dVal, err := val.ConvertTo(sc, x.ColsFieldType[i])
+				if err != nil {
+					return err
+				}
+				// The converted result must be same as original datum.
+				cmp, err := dVal.Compare(sc, &val, collate.GetCollator(x.ColsFieldType[i].Collate))
+				if err != nil || cmp != 0 {
+					return errors.New("The parameter for point get can not use plan cache, because the parameter has changed after the covert")
+				}
```

Comment:
How about warping this logic as a dedicated function like `convertConstantTo(con *Constant, target *Type) (*Datum, error)`?